rir/src/compiler/opt/early_constantfold.cpp

```diff
@@ -68,51 +70,130 @@ bool EarlyConstantfold::apply(Compiler&, ClosureVersion* cls, Code* code,
-                                                         PirType::any()));
-                                    a = *ip;
-                                    ip++;
-                                }
-                                if (a->type.isA(RType::prom) ||
-                                    a->type.maybePromiseWrapped()) {
-                                    ip = bb->insert(
-                                        ip, new Force(a, call->env(),
+                if (!nodots) {
+                    ip = next;
+                    continue;
+                }
+
+                if (auto cp = checkpoint.at(call)) {
+                    if (auto fun = Instruction::Cast(call->callArg(1).val())) {
+                        if (auto pc = fun->typeFeedback().feedbackOrigin.pc()) {
+                            auto bc = BC::decodeShallow(pc);
+                            assert(bc.bc == Opcode::record_type_);
+                            auto fb = bc.immediate.typeFeedback;
+                            auto seen = fb.seen.front();
+
+                            // only try if monomorphic callsite
+                            if (fb.numTypes != 1 ||
+                                (seen != SPECIALSXP && seen != BUILTINSXP &&
+                                 seen != CLOSXP)) {
+                                ip = next;
+                                continue;
+                            }
+
+                            anyChange = true;
+
+                            // speculate on the callee being special (no eager
+                            // evaluation), builtin (eagerly evaluate all
+                            // regardless of n), or closure (evaluate eagerly
+                            // according to n)
+                            ip = bb->insert(
+                                ip,
+                                new Is(
+                                    seen == SPECIALSXP
+                                        ? BC::RirTypecheck::isSPECIALSXP
+                                        : (seen == BUILTINSXP
+                                               ? BC::RirTypecheck::isBUILTINSXP
+                                               : BC::RirTypecheck::isCLOSXP),
+                                    fun));
+                            auto is = *ip;
+                            ip++;
+
+                            ip = bb->insert(ip,
+                                            new Identical(is, True::instance(),
+                                                          PirType::any()));
+                            auto test = *ip;
+                            ip++;
+
+                            ip = bb->insert(
+                                ip, new Assume(
+                                        test, cp,
+                                        DeoptReason(
+                                            fun->typeFeedback().feedbackOrigin,
+                                            DeoptReason::Typecheck)));
```

Comment:
I don't think this will work! the recordDeoptReason mechanism currently expects a typecheck to originate from a record_type_ kind of feedback. since this comes from call target recording I assume it will fail to update the typefeedback (or assert), leading to deopt loops. Or, did you actually test it?

Imho this is a callTarget deopt reason!